rsyslog/vendored_src/rsyslog/rsyslog/plugins/omelasticsearch/omelasticsearch.c

```diff
@@ -1482,6 +1482,12 @@ curlPost(wrkrInstanceData_t *pWrkrData, uchar *message, int msglen, uchar **tpls
 		/* by default, reuse existing connections */
 		curl_easy_setopt(curl, CURLOPT_FRESH_CONNECT, 0);
 	}
+	if ((pWrkrData->pData->rebindInterval > -1) &&
+		(pWrkrData->nOperations == pWrkrData->pData->rebindInterval)) {
+		curl_easy_setopt(curl, CURLOPT_FORBID_REUSE, 1);
+	} else {
+		curl_easy_setopt(curl, CURLOPT_FORBID_REUSE, 0);
+	}
```

Comment:
Is this just for testing purposes, or does this really solve the problem?
Why not put the `curl_easy_setopt(curl, CURLOPT_FORBID_REUSE, 1);` setting just after the `curl_easy_setopt(curl, CURLOPT_FRESH_CONNECT, 1);` setting and the `curl_easy_setopt(curl, CURLOPT_FORBID_REUSE, 0);` just after the `curl_easy_setopt(curl, CURLOPT_FRESH_CONNECT, 0);` setting?

Reply:
???  The patch works as expected, I believe.  Following is a snippet of Curl_done.  If reuse_forbid is set to true, Curl_disconnect is called.  So, I think we have to set it when we want to close the connection...

5648   if((data->set.reuse_forbid && !(conn->ntlm.state == NTLMSTATE_TYPE2 ||
5649                                   conn->proxyntlm.state == NTLMSTATE_TYPE2))
5650      || conn->bits.close || premature ||
5651      (-1 == conn->connection_id)) {
5652     CURLcode res2 = Curl_disconnect(conn, premature); /* close connection */
5653 
5654     /* If we had an error already, make sure we return that one. But
5655        if we got a new error, return that. */
5656     if(!result && res2)
5657       result = res2;
5658   }

Reply:
ok - please move `curl_easy_setopt(curl, CURLOPT_FORBID_REUSE, 1);` into the `if` clause with `curl_easy_setopt(curl, CURLOPT_FRESH_CONNECT, 1);`, and move `curl_easy_setopt(curl, CURLOPT_FORBID_REUSE, 0);` to the `else` clause with `curl_easy_setopt(curl, CURLOPT_FRESH_CONNECT, 0);`

Reply:
It may not be a big deal, but the closing connection timing would be changed by your proposal a little bit.  I wanted to close it before opening a new one in the previous cycle.  That keeps the count of ESTABLISHED connection always 1.  But if we move curl_easy_setopt(curl, CURLOPT_FORBID_REUSE, 1) next to curl_easy_setopt(curl, CURLOPT_FRESH_CONNECT, 1), you could have 2 connections for a short time.  If it's ok, I'll update the PR.

```
$ while true; do lsof -i -P -n |grep rsyslog; sleep 5; echo ""; done
rsyslogd  28431 nhosoi   58u  IPv4 3891135      0t0  TCP 10.195.132.86:38212->34.209.7.166:443 (ESTABLISHED)

rsyslogd  28431 nhosoi   58u  IPv4 3891135      0t0  TCP 10.195.132.86:38212->34.209.7.166:443 (ESTABLISHED)
rsyslogd  28431 nhosoi   60u  IPv4 3893709      0t0  TCP 10.195.132.86:38224->34.209.7.166:443 (ESTABLISHED)
```

Reply:
hmm - ok - I think it is better to close before opening a new one - if your patch already does that, then let's leave it as it is